src/main/java/com/amazon/pvar/tspoc/merlin/solver/flowfunctions/AbstractFlowFunctions.java

```diff
@@ -134,34 +138,29 @@ protected static NodeState makeNodeState(Node n) {
      * <p>
      * Classes that extend this abstract class should implement "visit" methods that update (but do not re-assign)
      * the nextStates attribute.
-     *
-     * @param node
-     * @return
      */
-    public final synchronized Set<State> computeNextStates(Node node, Value val) {
-        queryValue = val;
-        nextStates = new HashSet<>();
+    public final Set<State> computeNextStates() {
+        // If this assertion fails, we are inadvertently reusing the same flow function
+        // instance.
+        assert !transferApplied;
+        transferApplied = true;
```

Comment:
The bug I was mainly worried about is calling `computeNextStates` more than once on the same instance (from the same thread), which should no longer happen. You're right though that this would not catch this issue if called from two threads in an unlucky order. However, I think I don't want to introduce synchronization overhead purely for defensive programming.